src/Plutarch/Extra/Map.hs

```diff
@@ -1,185 +1,266 @@
+ Copyright: (C) Liqwid Labs 2022
+ License: Apache 2.0
+ Maintainer: Koz Ross <koz@mlabs.city>
+ Portability: GHC only
+ Stability: Experimental
+
+ Various helpers for 'PMap' use. This module is intended to be imported
+ qualified.
+-}
 module Plutarch.Extra.Map (
-    plookup,
-    plookup',
-    pmapFromList,
-    pkeys,
+    -- * Lookup
+    ptryLookup,
+
+    -- * Comparisons
+    pkeysEqual,
+
+    -- * Modification
     pupdate,
-    Plutarch.Extra.Map.pmap,
+
+    -- * Folds
+    pfoldMapWithKey,
+    pfoldlWithKey,
+
+    -- * Conversion
+    pmapFromFoldableUnsorted,
+    pmapFromFoldableSorted,
+    pkeys,
+
+    -- * Key-value pair manipulation
     pkvPairKey,
+    pkvPairValue,
     pkvPairLt,
-    pfoldlWithKey,
-    pfoldMapWithKey,
 ) where
 
 import Data.Foldable (foldl')
-import Plutarch.Api.V1.AssocMap (KeyGuarantees, PMap (PMap))
+import Plutarch.Api.V1.AssocMap (
+    KeyGuarantees (Sorted, Unsorted),
+    PMap (PMap),
+    pdelete,
+    pempty,
+    pinsert,
+    plookup,
+ )
 import Plutarch.Builtin (ppairDataBuiltin)
-import Plutarch.Extra.Functor (pfmap)
-import Plutarch.Extra.List (pmapMaybe)
-import Plutarch.Extra.TermCont (pletC, pmatchC)
-import Plutarch.Extra.Traversable (pfoldMap)
-import qualified Plutarch.List
-
--- | @since 1.0.0
-plookup ::
-    forall (k :: S -> Type) (v :: S -> Type) (keys :: KeyGuarantees) (s :: S).
-    (PIsData v, PIsData k, PEq k) =>
-    Term s (k :--> PMap keys k v :--> PMaybe v)
-plookup = phoistAcyclic $
-    plam $ \x m -> unTermCont $ do
-        PMap kvs <- pmatchC m
-        res <- pmatchC (pfind # (go # x) # kvs)
-        pure $ case res of
-            PNothing -> pcon PNothing
-            PJust y -> pcon . PJust . pfromData $ psndBuiltin # y
+import Plutarch.List (puncons)
+
+{- | Gives 'PTrue' if both argument 'PMap's contain mappings for exactly the
+ same set of keys. Requires a number of equality comparisons between keys
+ proportional to the length of the shorter argument.
+
+ @since 1.1.0
+-}
+pkeysEqual ::
+    forall (k :: S -> Type) (a :: S -> Type) (b :: S -> Type) (s :: S).
+    (PIsData k, PEq k) =>
+    Term s (PMap 'Sorted k a :--> PMap 'Sorted k b :--> PBool)
+pkeysEqual = phoistAcyclic $
+    plam $ \kvs kvs' ->
+        pmatch kvs $ \case
+            PMap ell -> pmatch kvs' $ \case
+                PMap ell' -> go # ell # ell'
   where
     go ::
         forall (s' :: S).
-        Term s' (k :--> PBuiltinPair (PAsData k) (PAsData v) :--> PBool)
+        Term
+            s'
+            ( PBuiltinList (PBuiltinPair (PAsData k) (PAsData a))
+                :--> PBuiltinList (PBuiltinPair (PAsData k) (PAsData b))
+                :--> PBool
+            )
     go = phoistAcyclic $
-        plam $ \x p -> unTermCont $ do
-            k' <- pletC (pfromData $ pfstBuiltin # p)
-            pure $ k' #== x
+        pfix #$ plam $ \self ell ell' ->
+            pmatch (puncons # ell) $ \case
+                PNothing -> pmatch (puncons # ell') $ \case
```

Comment:
I really think `pfromMaybe` on `Plutarch.Extra.Maybe` would make this code more terse. Personal opinion.

Reply:
ehh. maybe not because it's *very* nested.